[api-minor] Update the minimum supported Node.js version to 18

Snuffleupagus opened the pull request:
- **[api-minor] Update the minimum supported Node.js version to 18**

   This patch updates the minimum supported environments as follows:
    - Node.js 18, which was released on 2022-04-19; see https://en.wikipedia.org/wiki/Node.js#Releases

   Note also that Node.js 16 will soon reach EOL, and thus no longer receive any security updates.

 - **Remove the compatibility checks in `WorkerMessageHandler.createDocumentHandler`**

   For some time these checks have only targeted Node.js environments, since the features in question exist in all supported browsers (even when a `legacy`-build is used).

   Now that we've updated the minimum supported Node.js version to 18, a number of polyfills are thus (finally) no longer necessary in that environment. Hence for certain *basic* functionality, such as e.g. text-extraction, it's now possible to use either a modern- or a `legacy`-build of the PDF.js library in Node.js environments.

   *Please note:* For e.g. canvas-rendering in Node.js environments it's still necessary to use a `legacy`-build, since that functionality requires various polyfills.

 - **Remove the IMAGE_DECODERS special-case when polyfilling `structuredClone`**

   Originally we only used the `structuredClone` polyfill in the `LoopbackPort`-implementation, and that obviously isn't used anywhere within the various image decoders.
   At this point in time we've started to use `structuredClone` a little bit more, hence it seems overall simpler to just bundle the polyfill even in the `legacy`-version of the IMAGE_DECODERS built-target.

Snuffleupagus committed:
[api-minor] Update the minimum supported Node.js version to 18

This patch updates the minimum supported environments as follows:
 - Node.js 18, which was released on 2022-04-19; see https://en.wikipedia.org/wiki/Node.js#Releases

Note also that Node.js 16 will soon reach EOL, and thus no longer receive any security updates.

Snuffleupagus committed:
Remove the compatibility checks in `WorkerMessageHandler.createDocumentHandler`

For some time these checks have only targeted Node.js environments, since the features in question exist in all supported browsers (even when a `legacy`-build is used).

Now that we've updated the minimum supported Node.js version to 18, a number of polyfills are thus (finally) no longer necessary in that environment. Hence for certain *basic* functionality, such as e.g. text-extraction, it's now possible to use either a modern- or a `legacy`-build of the PDF.js library in Node.js environments.

*Please note:* For e.g. canvas-rendering in Node.js environments it's still necessary to use a `legacy`-build, since that functionality requires various polyfills.

timvandermeij commented:
I'm in favor of doing this considering that Node 18 is the current LTS version (and for e.g. Firefox we also only support the current ESR version), the upcoming EOL date of Node 16 and the fact that we already updated to Node 18 in the CI as well some time ago in https://github.com/mozilla/pdf.js/commit/588447aa908fadbe1cf4f9c892daa4b74da9dcca and didn't run into any issues. Moreover, it helps to remove compatibility code and a dependency, which is always good.

Snuffleupagus commented:
My plan was actually to wait a little bit before submitting this for actual review, since we *just* increased the minimum supported Node.js version to `16` in the latest release. Unfortunately some recent patches, see e.g. PR #16175, caused various regressions and it'd thus be good to have *at least one* official release that works in Node.js `16` before support is removed.

> the upcoming EOL date of Node 16 and the fact that we already updated to Node 18 in the CI as well some time ago in [588447a](https://github.com/mozilla/pdf.js/commit/588447aa908fadbe1cf4f9c892daa4b74da9dcca) and didn't run into any issues.

Part of the problem, w.r.t. the mentioned regressions, is that we only test in the Node.js `18` (i.e. the current LTS version) since that allowed the regressions to sneak in. From a PDF.js project perspective it doesn't feel great having to also test in old Node.js versions, and increasing the minimum support version would indeed considerably simplify maintenance.

Snuffleupagus committed:
Remove the IMAGE_DECODERS special-case when polyfilling `structuredClone`

Originally we only used the `structuredClone` polyfill in the `LoopbackPort`-implementation, and that obviously isn't used anywhere within the various image decoders.
At this point in time we've started to use `structuredClone` a little bit more, hence it seems overall simpler to just bundle the polyfill even in the `legacy`-version of the IMAGE_DECODERS built-target.

Snuffleupagus committed:
Enable `unicorn/prefer-at` unconditionally (PR 15014 follow-up)

Now that Node.js version 18 is required, we should be able to use `Array.prototype.at()` everywhere in the code-base.

Snuffleupagus commented:
/botio-linux preview

Snuffleupagus commented:
/botio unittest

timvandermeij commented:
Nice work; thanks!
